### PR TITLE
Add wired off code for Validation rules for Custom Resource Definitions using the CEL expression language

### PR DESCRIPTION
**Pull request by jpbetz**
This adds the dependencies and third_party/forked code import that is needed for https://github.com/kubernetes/kubernetes/pull/106051.

This PR purely intended to simplify the code review of https://github.com/kubernetes/kubernetes/pull/106051.  Once this PR merges, https://github.com/kubernetes/kubernetes/pull/106051 will have about 6k lines to review instead of 50k.

/kind feature
/sig api-machinery

```release-note
NONE
```

cc @cici37 @sttts @liggitt

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+This directory contains a port of https://github.com/google/cel-policy-templates-go/tree/master/policy/model modified in a few ways:
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/README.md:1]: include a copy of https://github.com/google/cel-policy-templates-go/blob/master/LICENSE at staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/README.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/README.md:2]: is this expected to be refreshed from `https://github.com/google/cel-policy-templates-go/tree/master/policy/model`? if so, should the deltas be made in their own commits so they can be replayed on top of refreshed copies re-imported later?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package validation
 
 import (
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go:20]: nit: group with below imports

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 4]
```diff
@@ -18,6 +18,14 @@ package validation
 
 import (
 	"encoding/json"
+	"fmt"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/ext"
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:26]: nit: standalone import group

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go [hunk 5]
```diff
@@ -18,6 +18,14 @@ package validation
...
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:28]: nit: group with k8s imports

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/README.md (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/README.md:2]: Good call. I've split out the delta.

**Comment by jpbetz** [comment]
Thanks for the review @liggitt, feedback applied.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go:20]: still outstanding?

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+// IsRootReserved returns true if an identifier is in the RootReservedIdentifiers set.
+func IsRootReserved(prop string) bool {
+	return RootReservedIdentifiers.Has(prop)
+}
+
+// Escape escapes identifiers in the AlwaysReservedIdentifiers set by prefixing ident with "_" and by prefixing
+// any ident already prefixed with N '_' with N+1 '_'.
+// For an identifier that does not require escaping, the identifier is returned as-is.
+func Escape(ident string) string {
+	if strings.HasPrefix(ident, "_") || AlwaysReservedIdentifiers.Has(ident) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping.go]: this turns `_foo` into `__foo`... is that necessary (and will that trigger the need to further escape to `__underscores__` as discussed in https://github.com/kubernetes/enhancements/pull/3039)? I would have thought we would only need to do that for things like `_int`

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go [hunk 7]
```diff
@@ -18,146 +18,115 @@ import (
...
+			"strVal": {Name: "strVal", Type: StringType},
+			"intVal": {Name: "intVal", Type: IntType},
+		})
+	}
+	if s.XEmbeddedResource {
+		return NewObjectType("embedded", map[string]*DeclField{
+			"kind":       {Name: "kind", Type: StringType},
+			"apiVersion": {Name: "apiVersion", Type: StringType},
+			"metadata": {Name: "metadata", Type: NewObjectType("metadata", map[string]*DeclField{
+				"name":         {Name: "name", Type: StringType},
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: this is also not required

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go [hunk 8]
```diff
@@ -18,146 +18,115 @@ import (
...
+				"generateName": {Name: "generateName", Type: StringType, Required: false},
+			})},
+			"spec":   {Name: "spec", Type: DynType},
+			"status": {Name: "status", Type: DynType, Required: false},
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: embedded resource doesn't imply root spec/status properties, only apiVersion/kind/metadata.name/metadata.generateName

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go [hunk 9]
```diff
@@ -18,146 +18,115 @@ import (
...
+		})
+	}
+	if s.XPreserveUnknownFields {
+		return DynType
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: returning here omits type definitions we know about at this level (I can specify properties a, b, and c, AND set x-kubernetes-preserve-unknown-fields, and we'd want to stay informed about the type information for properties a, b, and c)

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping_test.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+		{ unescaped: "null_type", escaped: "null_type", reservedAtRoot: true },
+		{ unescaped: "type", escaped: "type", reservedAtRoot: true },
+		// To prevent escaping from colliding with other identifiers, all identifiers prefixed by _s are escaped by
+		// prefixing them with N+1 _s.
+		{ unescaped: "_if", escaped: "__if" },
+		{ unescaped: "__if", escaped: "___if" },
+		{ unescaped: "___if", escaped: "____if" },
+		{ unescaped: "_has", escaped: "__has" },
+		{ unescaped: "_int", escaped: "__int" },
+		{ unescaped: "_anything", escaped: "__anything" },
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping_test.go:72]: this case surprised me... I wouldn't expect us to need to modify a single leading `_` if the non-`_` bits don't intersect with reserved identifiers

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+	// intOrStringType represents the x-kubernetes-int-or-string union type in CEL expressions.
+	// In CEL, the type is represented as an object where either the srtVal
+	// or intVal field is set. In CEL, this allows for typesafe expressions like:
+	//
+	// require that the string representation be a percentage:
+	//  `has(intOrStringField.strVal) && intOrStringField.strVal.matches(r'(\d+(\.\d+)?%)')`
+	// validate requirements on both the int and string representation:
+	//  `has(intOrStringField.intVal) ? intOrStringField.intVal < 5 : double(intOrStringField.strVal.replace('%', '')) < 0.5
+	//
+	intOrStringType = NewObjectType("intOrString", map[string]*DeclField{
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go:125]: needs a TODO to reconcile with updates in https://github.com/kubernetes/enhancements/pull/3039

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go (hunk 3 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation_test.go:20]: oops, fixed

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping.go (hunk 6 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping.go]: Added a TODO to swap in new escaping approach

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping_test.go (hunk 10 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/escaping_test.go:72]: Yes, we don't do that in the new approach, we only escape keywords and we do `{keyword}` -> `__{keyword}__`

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go (hunk 9 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: Fixed in later commits.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go (hunk 8 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: Fixed in later commits.

File: staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go (hunk 7 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go]: added a TODO to fix after merge

**Comment by liggitt** [comment]
```diff
 {
-	"directDependencies": 163,
-	"transitiveDependencies": 300,
-	"totalDependencies": 359,
-	"maxDepthOfDependencies": 29
+	"directDependencies": 164,
+	"transitiveDependencies": 306,
+	"totalDependencies": 366,
+	"maxDepthOfDependencies": 30
 }
```

* depth and direct dep counts bump by 1
* total deps bump by 7
* actual additional vendored deps: 3
  * [github.com/google/cel-go](https://github.com/google/cel-go) (expected, primary library supporting this feature)
  * [github.com/antlr/antlr4/tree/master/runtime/Go/antlr](https://github.com/antlr/antlr4/tree/master/runtime/Go/antlr) ([no transitive deps](https://github.com/antlr/antlr4/blob/master/runtime/Go/antlr/go.mod))
  * [github.com/stoewer/go-strcase](https://github.com/stoewer/go-strcase) ([single transitive dep](https://github.com/stoewer/go-strcase/blob/master/go.mod) we already had)

**Pull request merged**